### PR TITLE
plugin/rewrite: add EDNS0 revert regression tests for the record-present and replace paths (#8234)

### PR DESCRIPTION
**Pull request by maximilize**
### 1. Why is this pull request needed and what does it do?

Adds two regression tests for #8234. The production fix this PR originally carried is gone: #8235 landed on master and fixes the leak there, so only the missing test coverage remains.

The existing `TestRewriteEDNS0RevertDoesNotLeakThroughScrubWriter` drives an upstream reply with no records at all. These two cover paths it does not:

- `TestEdns0LocalSetRevertNoLeak` — an OPT-less reply that *does* carry a record, so the per-record response rules run while `ScrubWriter.SizeAndDo` reuses the request OPT.
- `TestEdns0LocalSetRevertRestoresPreexisting` — the replace path: the client request already carries option `0xffee`, so `set ... revert` must put the original value back rather than merely drop the option.

Both fail against the tree before #8235 (`56f7c0931~1`) and pass on current master. Verified with `gofmt -s`, `go vet`, `go test ./plugin/rewrite/ ./request/`, and `go test -race -count=20`.

### 2. Which issues (if any) are related?

#8234, tests for the fix merged in #8235.

### 3. Which documentation changes (if any) need to be made?

None.

### 4. Does this introduce a backward incompatible change or deprecation?

No — test-only.

### CODE REVIEW

File: plugin/rewrite/edns0.go [hunk 1]
```diff
@@ -42,9 +42,17 @@ type edns0NsidRule struct {
...
 	code uint16
+	// requestOpt is the request OPT the option was added to. It must be cleaned
+	// up on revert as well: the server's ScrubWriter.SizeAndDo reuses the request
+	// OPT when the reply has no OPT of its own, so an option left on the request
+	// leaks into the client reply (#8234).
+	requestOpt *dns.OPT
 }
 
 func (r *edns0SetResponseRule) RewriteResponse(res *dns.Msg, _ dns.RR) {
+	if r.requestOpt != nil {
```
Comment (houyuwushang) [inline plugin/rewrite/edns0.go]: This cleanup only runs when ResponseReverter.rewriteResourceRecord calls RewriteResponse. If the upstream writes a message with no Answer, Ns, or Extra records (for example, an empty SERVFAIL/NOERROR response), the loops in ResponseReverter.WriteMsg execute zero times, this block is skipped, and ScrubWriter still reuses the request OPT containing 0xffee. I reproduced that by using resp.SetReply(r) with no records; the final client response still contained 0xffee. Could the request restoration run once from the write path, independently of resource-record iteration, with a zero-RR regression case?

File: plugin/rewrite/edns0.go [hunk 2]
```diff
@@ -164,7 +172,7 @@ func (rule *edns0LocalRule) Rewrite(_ctx context.Context, state request.Request)
 	if rule.action == Append || rule.action == Set {
 		o.Option = append(o.Option, &dns.EDNS0_LOCAL{Code: rule.code, Data: rule.data})
 		if rule.revert {
-			resp = append(resp, &edns0SetResponseRule{code: rule.code})
+			resp = append(resp, &edns0SetResponseRule{code: rule.code, requestOpt: o})
```
Comment (houyuwushang) [inline plugin/rewrite/edns0.go]: The existing-option path is still uncovered. If the client request already has local option 0xffee=111111, set 0xffee abcdef revert takes the earlier match branch and creates edns0ReplaceResponseRule instead of this requestOpt-aware rule. With a no-OPT upstream response, that response rule cannot restore anything, and ScrubWriter reuses the request with abcdef; my probe received abcdef instead of the original 111111. The request-side restoration needs to cover replacement as well as addition.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
The added ScrubWriter regression test validates the reported answer-bearing path, and it passes reliably. However, request-side EDNS0 state is still not restored in two no-OPT response paths, so this does not yet fully close the leak:

1. A response with no Answer, Ns, or Extra records never invokes any ResponseRule.
2. A request that already contains the same option code uses edns0ReplaceResponseRule, which does not restore the mutated request option.

Please restore request-side state from a response-write hook that runs exactly once regardless of the response's RR count, and cover both cases.

Verification on 3017e586c:
- go test ./plugin/rewrite ./request: pass
- go test -race ./plugin/rewrite: pass
- go vet ./plugin/rewrite ./request: pass
- existing regression test with -count=20: pass
- targeted zero-RR response probe: fails, option 0xffee is returned
- targeted pre-existing-option probe: fails, original 111111 is returned as rewritten abcdef

**Comment by maximilize** [comment]
Both are real and now fixed — thanks for the precise repros, @houyuwushang.

You're right that the cleanup rode on `RewriteResponse`, which `ResponseReverter.WriteMsg` only drives per record, so an OPT-less reply with no records skipped it. I moved the request-OPT revert to a `revertRequest()` that `WriteMsg` runs once, independent of record count. The replace path (option already present on the request) now restores the original value in the request OPT too, instead of leaving the overwritten one.

Added regression tests for both probes: the empty-reply case, and the pre-existing `0xffee=111111` case (which now comes back as `111111`, not `abcdef`). `go test -race ./plugin/rewrite/` is green.

**Comment by maximilize** [comment]
@houyuwushang both of your probes have been covered since `616b4f5` — `revertRequest()` runs once from `WriteMsg` regardless of record count (zero-RR case), and the replace path restores the original request-side value; regression tests for both are in `edns0_test.go`. CI is green across the matrix and the branch still merges cleanly against master. Ready for another look whenever you have time.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
Thanks for the follow-up. The two previously reported cases are covered by the added tests and pass on `616b4f5`.

There is now a different blocker caused by the branch's old base: #8235 has since merged and fixes #8234 on master through `writeWithRevertedRequestExtra`. I applied only this PR's `edns0_test.go` to current master; all three tests pass, including 100 repeated runs. The production changes are therefore no longer needed.

When this head is merged with current master, `revertRequest()` runs first and permanently mutates the original request OPT before #8235 snapshots `currentExtra`. A focused object-state probe passes 100 times on master but fails on the merged tree: master preserves the upstream-rewritten `0xffee=abcdef` value on the request after `WriteMsg`, while the merged tree removes it. This defeats #8235's copy/defer preservation and leaves two overlapping request-revert mechanisms.

Please rebase onto current master, drop the production changes in `edns0.go` and `reverter.go`, and retain the useful missing test coverage for the zero-RR and pre-existing-option cases. The basic no-leak case is already covered by `TestRewriteEDNS0RevertDoesNotLeakThroughScrubWriter` on master.

I also ran the existing `plugin/rewrite` and `request` package tests, race tests, and `go vet` on the synthetic merge; they pass, so the existing suite does not catch this request-state regression.

**Comment by maximilize** [comment]
Rebased onto master, production changes in `edns0.go`/`reverter.go` dropped. The PR is now `plugin/rewrite/edns0_test.go` only, +120/−0.

One deviation: on master `TestRewriteEDNS0RevertDoesNotLeakThroughScrubWriter` already drives the zero-RR case — its handler replies with `SetReply(r)` and no records, and it fails against `56f7c0931~1`. So I dropped my duplicate of it and kept the two cases the suite genuinely misses:

- `TestEdns0LocalSetRevertNoLeak` — OPT-less reply that *does* carry a record, so the per-record response rules run while `SizeAndDo` reuses the request OPT.
- `TestEdns0LocalSetRevertRestoresPreexisting` — replace path: `0xffee=111111` on the request must come back as `111111`.

Both fail at `56f7c0931~1` and pass on current master. `gofmt -s`, `go vet ./plugin/rewrite/ ./request/`, `go test ./plugin/rewrite/ ./request/` and `go test -race -count=20` are green.

If you would rather have the zero-RR case stated explicitly here too, say so and I will re-add it.

**Review by houyuwushang — APPROVED** [review]
Re-reviewed 170d537. Rebasing and dropping the redundant production changes resolves my prior blocker. The two retained tests add distinct coverage for the answer-bearing ScrubWriter path and restoration of a pre-existing local option. I verified that each test fails on the parent of #8235 with the expected leak/value replacement and passes on the current head. No remaining blockers.
